// @ai-sdk/deepseek@1.0.30/dist/index.mjs
// src/deepseek-provider.ts
import { OpenAICompatibleChatLanguageModel } from "@ai-sdk/openai-compatible";
import {
  NoSuchModelError
} from "@ai-sdk/provider";
import {
  loadApiKey,
  withoutTrailingSlash,
  withUserAgentSuffix
} from "@ai-sdk/provider-utils";

// src/deepseek-metadata-extractor.ts
import { safeValidateTypes } from "@ai-sdk/provider-utils";
import { z } from "zod/v4";
var buildDeepseekMetadata = (usage) => {
  var _a, _b;
  return usage == null ? void 0 : {
    deepseek: {
      promptCacheHitTokens: (_a = usage.prompt_cache_hit_tokens) != null ? _a : NaN,
      promptCacheMissTokens: (_b = usage.prompt_cache_miss_tokens) != null ? _b : NaN
    }
  };
};
var deepSeekMetadataExtractor = {
  extractMetadata: async ({ parsedBody }) => {
    const parsed = await safeValidateTypes({
      value: parsedBody,
      schema: deepSeekResponseSchema
    });
    return !parsed.success || parsed.value.usage == null ? void 0 : buildDeepseekMetadata(parsed.value.usage);
  },
  createStreamExtractor: () => {
    let usage;
    return {
      processChunk: async (chunk) => {
        var _a, _b;
        const parsed = await safeValidateTypes({
          value: chunk,
          schema: deepSeekStreamChunkSchema
        });
        if (parsed.success && ((_b = (_a = parsed.value.choices) == null ? void 0 : _a[0]) == null ? void 0 : _b.finish_reason) === "stop" && parsed.value.usage) {
          usage = parsed.value.usage;
        }
      },
      buildMetadata: () => buildDeepseekMetadata(usage)
    };
  }
};
var deepSeekUsageSchema = z.object({
  prompt_cache_hit_tokens: z.number().nullish(),
  prompt_cache_miss_tokens: z.number().nullish()
});
var deepSeekResponseSchema = z.object({
  usage: deepSeekUsageSchema.nullish()
});
var deepSeekStreamChunkSchema = z.object({
  choices: z.array(
    z.object({
      finish_reason: z.string().nullish()
    })
  ).nullish(),
  usage: deepSeekUsageSchema.nullish()
});

// src/version.ts
var VERSION = true ? "1.0.30" : "0.0.0-test";

// src/deepseek-provider.ts
function createDeepSeek(options = {}) {
  var _a;
  const baseURL = withoutTrailingSlash(
    (_a = options.baseURL) != null ? _a : "https://api.deepseek.com/v1"
  );
  const getHeaders = () => withUserAgentSuffix(
    {
      Authorization: `Bearer ${loadApiKey({
        apiKey: options.apiKey,
        environmentVariableName: "DEEPSEEK_API_KEY",
        description: "DeepSeek API key"
      })}`,
      ...options.headers
    },
    `ai-sdk/deepseek/${VERSION}`
  );
  class DeepSeekChatLanguageModel extends OpenAICompatibleChatLanguageModel {
    addJsonInstruction(opts) {
      var _a2;
      if (((_a2 = opts.responseFormat) == null ? void 0 : _a2.type) !== "json") return opts;
      const promptArray = Array.isArray(opts.prompt) ? opts.prompt : [];
      const instruction = "Return ONLY a valid JSON object.";
      const adjustedPrompt = [
        ...promptArray,
        { role: "user", content: [{ type: "text", text: instruction }] }
      ];
      return { ...opts, prompt: adjustedPrompt };
    }
    async doGenerate(options2) {
      return super.doGenerate(this.addJsonInstruction(options2));
    }
    async doStream(options2) {
      return super.doStream(this.addJsonInstruction(options2));
    }
  }
  const createLanguageModel = (modelId) => {
    return new DeepSeekChatLanguageModel(modelId, {
      provider: `deepseek.chat`,
      url: ({ path }) => `${baseURL}${path}`,
      headers: getHeaders,
      fetch: options.fetch,
      metadataExtractor: deepSeekMetadataExtractor
    });
  };
  const provider = (modelId) => createLanguageModel(modelId);
  provider.languageModel = createLanguageModel;
  provider.chat = createLanguageModel;
  provider.textEmbeddingModel = (modelId) => {
    throw new NoSuchModelError({ modelId, modelType: "textEmbeddingModel" });
  };
  provider.imageModel = (modelId) => {
    throw new NoSuchModelError({ modelId, modelType: "imageModel" });
  };
  return provider;
}
var deepseek = createDeepSeek();
export {
  VERSION,
  createDeepSeek,
  deepseek
};
